salt/config/__init__.py

```diff
@@ -840,6 +840,13 @@ def _gather_buffer_space():
     # Always generate minion id in lowercase.
     'minion_id_lowercase': bool,
 
+    # The behavior when generating a minion ID. Can specify 'str' or 'func'.
+    # If 'str' is specified (the default), the 'id' option should be set to
+    # the minion ID or left unspecified for default minion ID generation
+    # behavior. If 'func' is specified, the 'id' option should be set to an
+    # exec module function to run to determine the minion ID.
+    'minion_id_type': str,
```

Comment:
@cachedout Hmm. I would rather go here in "general mechanism", like if you're need to first run some 3rd party Perl script that creates a file that you  will read and a bit parse — then you _do need_ a function. @skizunov only described a simple thing that will end-up an appendix code to the other things. And that part I don't like. If we are doing pluggable ID generator, let it be overall pluggable, even if this is just to read a plain text from somewhere.